gips/0065-subgraph-availability-manager.md

```diff
@@ -0,0 +1,125 @@
+---
+GIP: 0065
+Title: Subgraph Availability Manager
+Authors: Miguel de Elias <miguel@edgeandnode.com>
+Created: 2024-01-10
+Stage: Draft
+Discussions-To: https://forum.thegraph.com/t/gip-0065-subgraph-availability-manager/5501
+Category: Protocol Interfaces
+Implementations: https://github.com/graphprotocol/contracts/pull/882
+Audits: https://github.com/graphprotocol/contracts/blob/main/packages/contracts/audits/OpenZeppelin/2024-02-subgraph-availability-manager-and-minimum-allocation-duration-removal.pdf
+---
+
+# Abstract
+
+This GIP proposes increasing the decentralization in The Graph by introducing a new smart contract `SubgraphAvailabilityManager`.
+
+# Motivation
+
+The Subgraph Availability Oracle (SAO) plays a crucial role in The Graph's ecosystem. Its main function is to verify the availability and validity of subgraph files. Currently there is a single instance of the SAO in operation executing open-source code accessible at [The Graph's Subgraph Oracle repository](https://github.com/graphprotocol/subgraph-oracle). This SAO instance runs at five-minute intervals, performing a series of checks to assess the state of subgraph files. In instances where the SAO identifies unavailability or invalidity in a subgraph's files, it sends a transaction to the `RewardsManager` contract to deny indexing rewards for that subgraph. By default, subgraphs are presumed to be available, and typically, no initial action or vote by the SAO is necessary when a new subgraph is added to the network.
```

Comment:
With the current amount of subgraphs (1,000 on arb), how long does it take to process these checks? Are there any concerns with how this would operate with 100,000 subgraphs?

Reply:
Also, are all checks related to items contained in the manifest (features/networks). Are there any circumstances where features are used in the mappings but not included in the manifest? `ipfs.cat()` as an example

Reply:
You can find all checks it performs [here](https://github.com/graphprotocol/subgraph-oracle/blob/main/availability-oracle/src/main.rs#L415). Most checks are for items contained in manifest, there's no check for features used in mappings but not included in manifest but we do restrict certain host function calls from `ipfs.cat()`.  Also note that changes to SAO's code are out of scope for this GIP but the repository is open source so feel free to add comments there.

Reply:
> With the current amount of subgraphs (1,000 on arb), how long does it take to process these checks? Are there any concerns with how this would operate with 100,000 subgraphs?

Currently it takes 60 seconds to check all subgraphs. We might need a different approach for those numbers.